inc/sitemaps/class-post-type-sitemap-provider.php

```diff
@@ -0,0 +1,457 @@
+
+			$offset += $steps;
+
+			if ( empty( $posts ) ) {
+				continue;
+			}
+
+			$post_ids = wp_list_pluck( $posts, 'ID' );
+			$this->image_parser->cache_attachments( $post_ids );
+
+			$posts_to_exclude = explode( ',', $this->options['excluded-posts'] );
+
+			foreach ( $posts as $post ) {
+
+				if ( WPSEO_Meta::get_value( 'meta-robots-noindex', $post->ID ) === '1' ) {
+					continue;
+				}
+
+				if ( in_array( $post->ID, $posts_to_exclude ) ) {
+					continue;
+				}
+
+				$url = $this->get_url( $post );
+
+				if ( ! isset( $url['loc'] ) || in_array( $url['loc'], $stackedurls ) ) {
+					continue;
+				}
+
+				// Use this filter to adjust the entry before it gets added to the sitemap.
+				// TODO document filter. R.
+				$url = apply_filters( 'wpseo_sitemap_entry', $url, 'post', $post );
+
+				if ( ! empty( $url ) ) {
+					$links[]       = $url;
+					$stackedurls[] = $url['loc'];
+				}
+			}
+			unset( $post, $url );
+		}
+
+		return $links;
+	}
+
+	/**
+	 * Check if post type should be present in sitemaps.
+	 *
+	 * @param string $post_type Post type string to check for.
+	 *
+	 * @return bool
+	 */
+	public function is_valid_post_type( $post_type ) {
+
+		if ( ! empty( $this->options[ "post_types-{$post_type}-not_in_sitemap" ] ) ) {
+			return false;
+		}
+
+		if ( in_array( $post_type, array( 'revision', 'nav_menu_item' ) ) ) {
+			return false;
+		}
+
+		// TODO document filter. R.
+		if ( apply_filters( 'wpseo_sitemap_exclude_post_type', false, $post_type ) ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Get count of posts for post type.
+	 *
+	 * @param string $post_type Post type to retrieve count for.
+	 *
+	 * @return int
+	 */
+	protected function get_post_type_count( $post_type ) {
+
+		global $wpdb;
+
+		// TODO document filters. R.
+		$join_filter  = apply_filters( 'wpseo_typecount_join', '', $post_type );
+		$where_filter = apply_filters( 'wpseo_typecount_where', '', $post_type );
+
+		$sql   = "
+			SELECT COUNT(ID)
+			FROM {$wpdb->posts}
+			{$join_filter}
+			WHERE post_status IN ('publish','inherit')
+				AND post_password = ''
+				AND post_date != '0000-00-00 00:00:00'
+				AND post_type = %s
+				{$where_filter}
+		";
+
+		return (int) $wpdb->get_var( $wpdb->prepare( $sql, $post_type ) );
+	}
+
+	/**
+	 * Produces set of links to prepend at start of first sitemap page.
+	 *
+	 * @param string $post_type Post type to produce links for.
+	 *
+	 * @return array
+	 */
+	protected function get_first_links( $post_type ) {
+
+		$links = array();
+
+		$front_id = get_option( 'page_on_front' );
+
+		if ( ! $front_id && ( $post_type == 'post' || $post_type == 'page' ) ) {
+
+			$links[] = array(
+				'loc' => $this->home_url,
+				'pri' => 1,
+				'chf' => WPSEO_Sitemaps::filter_frequency( 'homepage', 'daily', $this->home_url ),
+			);
+		}
+		elseif ( $front_id && $post_type == 'post' ) {
+
+			$page_for_posts = get_option( 'page_for_posts' );
+
+			if ( $page_for_posts ) {
```

Comment:
Maybe you can combine this `if` and the `elseif` above, to keep the cyclomatic lower.
`elseif ( $front_id && $post_type == 'post' && $page_for_posts = get_option( 'page_for_posts' ) ) {`